validity-test-ospath/OsPath/Gen/Include.hs

```diff
@@ -0,0 +1,145 @@
+  shrinkValid = filter isValid . shrinkValidWith parseAbsFile
+
+instance GenValid (Path Abs Dir) where
+  genValid = (Path . ([OsString.pstr|/|] <>) . (<> OsString.singleton OsPath.pathSeparator) <$> genValid) `suchThat` isValid
+  shrinkValid = filter isValid . shrinkValidWith parseAbsDir
+
+instance GenValid (Path Rel File) where
+  genValid = (Path <$> genValid) `suchThat` isValid
+  shrinkValid = filter isValid . shrinkValidWith parseRelFile
+
+instance GenValid (Path Rel Dir) where
+  genValid = (Path . (<> OsString.singleton OsPath.pathSeparator) <$> genValid) `suchThat` isValid
+  shrinkValid = filter isValid . shrinkValidWith parseRelDir
+
+instance GenValid (SomeBase Dir) where
+  genValid = genValidStructurallyWithoutExtraChecking
+  shrinkValid = shrinkValidStructurallyWithoutExtraFiltering
+
+instance GenValid (SomeBase File) where
+  genValid = genValidStructurallyWithoutExtraChecking
+  shrinkValid = shrinkValidStructurallyWithoutExtraFiltering
+
+validateCommon :: Path b t -> Validation
+validateCommon (Path fp) = mconcat
+  [ declare "System.FilePath considers the path valid if it's not empty." $
+      OsPath.isValid fp || OsString.null fp
+  , declare "The path does not contain a '..' path component." $
+      not (hasParentDir fp)
+  ]
+
+validateDirectory :: Path b Dir -> Validation
+validateDirectory (Path fp) = mconcat
+  [ declare "The path has a trailing path separator if it's not empty." $
+      OsPath.hasTrailingPathSeparator fp || OsString.null fp
+  ]
+
+validateFile :: Path b File -> Validation
+validateFile (Path fp) = mconcat
+  [ declare "The path has no trailing path separator." $
+      not (OsPath.hasTrailingPathSeparator fp)
+  , declare "The path does not equal \".\"" $
+      fp /= [OsString.pstr|.|]
+  , declare "The path does not end in /." $
+      not ([OsString.pstr|/.|] `OsString.isSuffixOf` fp)
+  ]
+
+validateAbs :: Path Abs t -> Validation
+validateAbs (Path fp) = mconcat
+  [ declare "The path is absolute." $
+      OsPath.isAbsolute fp
+  ]
+
+validateRel :: Path Rel t -> Validation
+validateRel (Path fp) = mconcat
+  [ declare "The path is relative." $
+      OsPath.isRelative fp
+  ]
+
+shrinkValidWith :: (PLATFORM_PATH -> Maybe (Path a b)) -> Path a b -> [Path a b]
+shrinkValidWith fun (Path f) = filter (/= Path f) . mapMaybe fun $ shrinkValid f
+
+--------------------------------------------------------------------------------
+-- Orphan instances
+
+-- | Generates 'PLATFORM_PATH with a high occurence of @'.'@, @'\/'@ and
+-- @'\\'@ characters. The resulting 'FilePath's are not guaranteed to
+-- be valid.
+instance GenValid PLATFORM_PATH where
+    -- We also need to exclude UTF-16 surrogates.
+    genValid = mconcat <$> listOf (OsString.unsafeEncodeUtf . (:[]) . chr <$> frequency
+        [ (2, choose (0x0, 0xD800 - 1))
+        , (2, choose (0xDFFF + 1, 0x10FFFF))
+        , (1, elements (map ord "./\\"))
+        ]
+        )
+    shrinkValid _ = [] -- TODO: Not yet implemented
```

Comment:
Didn't we have shrinking before?
This will be important if/when we run into bugs but it's only a maintenance burden so maybe not so important for users.

Reply:
Yes, we did; But the one for `FilePath` is (obviously) the the same as the one for the general-purpose `String`, which is provided by the `genvalidity` package. Since there are no pre-defined instances for `OsPath`, `OsString` or `OsChar` and we have to roll our own, I thought that we might reconsider how we shrink paths specifically.
I didn't find the time to do that, but I didn't want to blindly copy existing code either; Hence the TODO...
But you are right, before we merge this I want to provide an implementation one way or the other.

Reply:
Revisiting the `genValid` implementation as well, this looks wrong to me as well: IMHO the generator should not only produce Unicode paths, but random sequences of `Word8`/`Word16` separated by `/`, `\` and `.`.
I.e. we should not assume an encoding.

Reply:
Writing a meaningf generator for windows paths is incredibly hard.

`filepath` has done this by expressing them in ABNF.

See here: https://github.com/haskell/filepath/blob/master/tests/filepath-equivalent-tests/Gen.hs

Reply:
Uh, that looks horrifying! Thanks for the pointer though.

Reply:
> Uh, that looks horrifying! Thanks for the pointer though.

It is beautiful, because it's a pretty good ABNF. I'm planning to actually represent filepaths through that data structure at some point. Otherwise it's not possible to have correct splitting etc. Lots of functions in filepath are ad-hoc.